src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.Subscription.cs

```diff
@@ -0,0 +1,182 @@
+        /// </para>
+        /// <para>
+        /// Initialization of the Dataflow subscription happens lazily, upon the first up-to-date check request.
+        /// </para>
+        /// <para>
+        /// Destruction of the Dataflow subscription happens when the parent component is disposed or unloaded.
+        /// </para>
+        /// </remarks>
+        private sealed class Subscription : IDisposable
+        {
+            /// <summary>
+            /// Completes when the first project update is received. Cancelled if the subscription is disposed.
+            /// </summary>
+            /// <remarks>
+            /// This field is also used to synchronise updates to <see cref="_link"/> and <see cref="State"/>.
+            /// </remarks>
+            private readonly TaskCompletionSource<byte> _dataReceived = new();
+
+            /// <summary>
+            /// Prevent overlapping requests.
+            /// </summary>
+            private readonly AsyncSemaphore _semaphore = new(1);
+
+            /// <summary>
+            /// Current <see cref="BuildUpToDateCheck.State"/> of the instance.
+            /// </summary>
+            /// <remarks>
+            /// Internal to support unit testing only.
+            /// </remarks>
+            internal State State { get; set; } = State.Empty;
+
+            /// <summary>
+            /// Lazily constructed Dataflow subscription. Set back to <see langword="null"/> in <see cref="Dispose"/>.
+            /// </summary>
+            private IDisposable? _link;
+
+            /// <summary>
+            /// Cancelled when this instance is disposed.
+            /// </summary>
+            private readonly CancellationTokenSource _disposeTokenSource = new();
+
+            public async Task<bool> RunAsync(
+                Func<State, CancellationToken, Task<bool>> func,
+                ConfiguredProject configuredProject,
+                IProjectItemSchemaService projectItemSchemaService,
+                CancellationToken cancellationToken)
+            {
+                using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _disposeTokenSource.Token);
+
+                CancellationToken token = cts.Token;
+
+                // Throws if the subscription has been disposed, or the caller's token cancelled.
+                token.ThrowIfCancellationRequested();
+
+                // Note that we defer subscription until an actual request is made in order to
+                // prevent redundant work/allocation for inactive project configurations.
+                // https://github.com/dotnet/project-system/issues/6327
+                //
+                // We don't pass the cancellation token here as initialization must be atomic.
+                EnsureInitialized();
+
+                token.ThrowIfCancellationRequested();
+
+                // Wait for the first state to be computed
+                await _dataReceived.Task.WithCancellation(token);
```

Comment:
Do we not do `ConfigureAwait(false)` on all async calls in this repo? I thought any library used for application purposes (those that have UI threads) should always have `ConfigureAwait(false)`.

Reply:
That is common advice, but we do not do so in this repo. Nor does CPS.

Reply:
ConfigureAwait(false) says "schedule the continuation on a thread-pool thread". This is mostly not appropriate in VS and should only be done deliberately.